### PR TITLE
resolved: just the dnssec bypass logic

### PR DESCRIPTION
**Pull request by poettering**
This is the first part of #17535, the actual code for the DNSSEC bypass logic. The other commits in that mega PR are further fixes, but let's focus on the main work first.

### CODE REVIEW

File: src/resolve/resolved-dns-stub.c [hunk 1]
```diff
@@ -703,13 +703,13 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
...
                 log_debug("Got message with obsolete key type, refusing.");
-                dns_stub_send_failure(m, l, s, p, DNS_RCODE_NOTIMP, false);
+                dns_stub_send_failure(m, l, s, p, DNS_RCODE_FORMERR, false);
                 return;
         }
 
         if (dns_type_is_zone_transer(p->question->keys[0]->type)) {
                 log_debug("Got request for zone transfer, refusing.");
-                dns_stub_send_failure(m, l, s, p, DNS_RCODE_NOTIMP, false);
+                dns_stub_send_failure(m, l, s, p, DNS_RCODE_FORMERR, false);
```
Comment (pemensik) [inline src/resolve/resolved-dns-stub.c]: Why is FORMERR sent for AXFR requests? Just return REFUSED. It is a valid query, it is understood what request is, but refusing to comply. I think REFUSED is more appropriate.

File: src/resolve/resolved-dns-stub.c [hunk 2]
```diff
@@ -703,13 +703,13 @@ static void dns_stub_process_query(Manager *m, DnsStubListenerExtra *l, DnsStrea
 
         if (dns_type_is_obsolete(p->question->keys[0]->type)) {
                 log_debug("Got message with obsolete key type, refusing.");
-                dns_stub_send_failure(m, l, s, p, DNS_RCODE_NOTIMP, false);
+                dns_stub_send_failure(m, l, s, p, DNS_RCODE_FORMERR, false);
```
Comment (pemensik) [inline src/resolve/resolved-dns-stub.c]: Why don't you send REFUSED, when logged message is refusing? I am not sure what is reason for refusing those requests. Any of them does not seem invalid, they are harmless anyway. Is there reason to refuse them explicitly?

Comment (poettering) [reply src/resolve/resolved-dns-stub.c]: Paranoia is a good reason to me.

Comment (pemensik) [reply src/resolve/resolved-dns-stub.c]: Are they dangerous in any sense?

File: src/resolve/resolved-varlink.c [hunk 3]
```diff
@@ -103,7 +106,7 @@ static bool validate_and_mangle_flags(
         /* This checks that the specified client-provided flags parameter actually makes sense, and mangles
          * it slightly. Specifically:
          *
-         * 1. We check that only the protocol flags and the NO_CNAME flag are on at most, plus the
+         * 1. We check that only the protocol flags and a nunch of NO_XYZ flags are on at most, plus the
```
Comment (keszybz) [inline src/resolve/resolved-varlink.c]: nunch → bunch, I guess

File: src/resolve/resolved-dns-stub.c (hunk 2 above)
Comment (keszybz) [reply src/resolve/resolved-dns-stub.c]: Yeah, FORMERR seems less even appropriate than NOTIMP. It makes things even more confusing.

**Review by keszybz — COMMENTED** [review]
Something is broken with "resolved: add support for answering DNSSEC questions on the stub". I don't see what exactly, but I bisect says that that is the first bad commit:
```
$ resolvectl query mirrors.fedoraproject.org
mirrors.fedoraproject.org: 2a05:d01c:c6a:cc03:263a:8409:b961:7e02
                           2620:52:3:1:dead:beef:cafe:fed7
                           2605:bc80:3010:600:dead:beef:cafe:fed9
                           2620:52:3:1:dead:beef:cafe:fed6
                           2001:4178:2:1269::fed2
                           2604:1580:fe00:0:dead:beef:cafe:fed1
                           2610:28:3090:3001:dead:beef:cafe:fed3
                           2a05:d014:10:7803:f774:4d7c:e277:a457
                           152.19.134.198
                           140.211.169.206
                           209.132.190.2
                           18.185.136.17
                           38.145.60.21
                           67.219.144.68
                           185.141.165.254
                           18.159.254.57
                           8.43.85.67
                           38.145.60.20
                           152.19.134.142
                           18.130.159.183
                           85.236.55.6
                           (wildcard.fedoraproject.org)

-- Information acquired via protocol DNS in 2.9ms.
-- Data is authenticated: no
$ ping mirrors.fedoraproject.org
ping: mirrors.fedoraproject.org: No address associated with hostname
$ getent hosts mirrors.fedoraproject.org
```
This causes the `dnf` failures that I mentioned previously.

**Comment by keszybz** [comment]
Success:
```json
systemd-resolved[22406]: varlink-23: Sending message: {"parameters":{"addresses":[{"ifindex":2,"family":2,"address":[18,185,136,17]},{"ifindex":2,"family":2,"address":[152,19,134,198]},{"ifindex":2,"family":2,"address":[185,141,165,254]},{"ifindex":2,"family":2,"address":[140,211,169,206]},{"ifindex":2,"family":2,"address":[85,236,55,6]},{"ifindex":2,"family":2,"address":[67,219,144,68]},{"ifindex":2,"family":2,"address":[152,19,134,142]},{"ifindex":2,"family":2,"address":[38,145,60,21]},{"ifindex":2,"family":2,"address":[8,43,85,67]},{"ifindex":2,"family":2,"address":[209,132,190,2]},{"ifindex":2,"family":2,"address":[18,130,159,183]},{"ifindex":2,"family":2,"address":[38,145,60,20]},{"ifindex":2,"family":2,"address":[18,159,254,57]}],"name":"wildcard.fedoraproject.org","flags":1}}
```
Failure:
```json
systemd-resolved[22406]: varlink-23: Sending message: {"parameters":{"addresses":[{"ifindex":0,"family":2,"address":[38,145,60,20]},{"ifindex":0,"family":2,"address":[18,185,136,17]},{"ifindex":0,"family":2,"address":[209,132,190,2]},{"ifindex":0,"family":2,"address":[38,145,60,21]},{"ifindex":0,"family":2,"address":[140,211,169,206]},{"ifindex":0,"family":2,"address":[85,236,55,6]},{"ifindex":0,"family":2,"address":[18,159,254,57]},{"ifindex":0,"family":2,"address":[18,130,159,183]},{"ifindex":0,"family":2,"address":[8,43,85,67]},{"ifindex":0,"family":2,"address":[185,141,165,254]},{"ifindex":0,"family":2,"address":[67,219,144,68]},{"ifindex":0,"family":2,"address":[152,19,134,198]},{"ifindex":0,"family":2,"address":[152,19,134,142]}],"name":"wildcard.fedoraproject.org","flags":1}}
```
ifindex==0 is rejected by json_dispatch_ifindex() in nss-resolve.c.

**Comment by keszybz** [comment]
@poettering any progress on this?

**Comment by poettering** [comment]
@keszybz what's your "resolvectl" output when this happens, btw? i.e. i take it you have only system wide DNS settings, not per-interface onces, i.e. no NM?

**Comment by keszybz** [comment]
> @keszybz what's your "resolvectl" output when this happens, btw? i.e. i take it you have only system wide DNS settings, not per-interface onces, i.e. no NM?

```console
$ resolvectl dns
Global:
Link 2 (enp1s0): 192.168.1.254 fe80::267f:20ff:fe41:bf6d%32610
$ networkctl 
IDX LINK   TYPE     OPERATIONAL SETUP     
  1 lo     loopback carrier     unmanaged 
  2 enp1s0 ether    routable    configured

2 links listed.
```
I think this '2' for 'enp1s0' is where the '2' in https://github.com/systemd/systemd/pull/17823#issuecomment-742439422 was from.

**Comment by poettering** [comment]
hmm, this is weird, i wonder where the ifindex info got lost then. can you paste a debug log excerpt from a case where it fails, i.e. the ifindex went missing?

**Comment by poettering** [comment]
(I prepped #18482 now, which will make the issue go away, but doesn't actually fix it (though we still should apply it). i really wonder how the ifindex info got lost in your case)

**Comment by keszybz** [comment]
> hmm, this is weird, i wonder where the ifindex info got lost then. can you paste a debug log excerpt from a case where it fails, i.e. the ifindex went missing?

I'll try on Monday. I don't have this version on the code on that VM anymore.

**Comment by poettering** [comment]
OK, I think I found it. Should work now.  I mean, I know where we lost the ifindex, and I fixed taht, but I don't actually know how to best reproduce the issue...

Hence, please give this a whirl!

**Comment by poettering** [comment]
and when you test this on top of #18483 and/or #18482, please note that the failing/non-failing "ping" cannot be used as verification, as those two PRs hide the fact that the ifindex got lost. A good way to verify is your "resolvectl query" command however. If the ifindex is known you should see a "comment" column about that, i.e. it should look like this:

```c
$ resolvectl query mirrors.fedoraproject.org
mirrors.fedoraproject.org: 85.236.55.6         -- link: wlp2s0
                           185.141.165.254     -- link: wlp2s0
                           209.132.190.2       -- link: wlp2s0
                           8.43.85.67          -- link: wlp2s0
                           18.133.140.134      -- link: wlp2s0
                           38.145.60.20        -- link: wlp2s0
                           67.219.144.68       -- link: wlp2s0
                           38.145.60.21        -- link: wlp2s0
                           18.185.136.17       -- link: wlp2s0
                           152.19.134.142      -- link: wlp2s0
                           152.19.134.198      -- link: wlp2s0
                           18.159.254.57       -- link: wlp2s0
                           140.211.169.206     -- link: wlp2s0
                           2a05:d01c:c6a:cc01:269:da52:9ae1:43e6 -- link: wlp2s0
                           2605:bc80:3010:600:dead:beef:cafe:fed9 -- link: wlp2s0
                           2a05:d014:10:7803:f774:4d7c:e277:a457 -- link: wlp2s0
                           2620:52:3:1:dead:beef:cafe:fed6 -- link: wlp2s0
                           2610:28:3090:3001:dead:beef:cafe:fed3 -- link: wlp2s0
                           2001:4178:2:1269::fed2 -- link: wlp2s0
                           2620:52:3:1:dead:beef:cafe:fed7 -- link: wlp2s0
                           2604:1580:fe00:0:dead:beef:cafe:fed1 -- link: wlp2s0
                           (wildcard.fedoraproject.org)

-- Information acquired via protocol DNS in 20.1ms.
-- Data is authenticated: no
```

if the ifindex got lost all the `-- link:` stuff on the right will be missing. if it's missing the bug remains unfixed, otherwise I think it's safe to say i managed to fix it.

**Comment by keszybz** [comment]
Rebase please.

**Comment by poettering** [comment]
(sniff, but it was all so fully green this time! i was so hoping to get a fully green PR merged for once ;-))

**Comment by poettering** [comment]
rebased. no other changes

**Comment by poettering** [comment]
yay, it passed again all green! yippieh!

File: src/resolve/resolved-dns-cache.c [hunk 4]
```diff
@@ -50,6 +53,12 @@ struct DnsCacheItem {
         LIST_FIELDS(DnsCacheItem, by_key);
 };
 
+/* Returns true if this is a cache item created as result of an explicit lookup, or created as "side-effect"
+ * of another request. "Primary" entries will carry the full answer data (with NSEC, …) that can aso prove
```
Comment (keszybz) [inline src/resolve/resolved-dns-cache.c:57]: also

File: src/resolve/resolved-dns-cache.c [hunk 5]
```diff
@@ -685,18 +726,49 @@ int dns_cache_put(
...
+                int primary = false;
+
+                if (!FLAGS_SET(item->flags, DNS_ANSWER_CACHEABLE) ||
                     !rr_eligible(item->rr))
                         continue;
 
+                if (key) {
+                        /* We store the auxiliary RRs and packet data in the cache only if they were in
+                         * direct response to the original query. If we cache an RR we also received, and
+                         * that is just auxiliary information we can't use the data, hence don't. */
```
Comment (keszybz) [inline src/resolve/resolved-dns-cache.c:737]: Something's wrong with this sentence.

File: src/resolve/resolved-dns-stub.c [hunk 6]
```diff
@@ -85,69 +85,259 @@ DnsStubListenerExtra *dns_stub_listener_extra_free(DnsStubListenerExtra *p) {
...
+
+                if (dns_answer_contains(exclude1, item->rr) ||
+                    dns_answer_contains(exclude2, item->rr))
+                        continue;
+
+                if (!with_dnssec &&
+                    dns_type_is_dnssec(item->rr->key->type))
+                        continue;
+
+                if (((item->flags ^ section) & (DNS_ANSWER_SECTION_ANSWER|DNS_ANSWER_SECTION_AUTHORITY|DNS_ANSWER_SECTION_ADDITIONAL)) != 0)
```
Comment (keszybz) [inline src/resolve/resolved-dns-stub.c:164]: Maybe add a comment explaining what the rule here is... This isn't obvious.

File: src/resolve/resolved-dns-stub.c [hunk 7]
```diff
@@ -171,19 +362,21 @@ static int dns_stub_finish_reply_packet(
...
                         rcode = DNS_RCODE_SERVFAIL;
         }
 
-        /* Don't set the AD bit unless DO is on, too */
-        if (!edns0_do)
+        /* Don't set the AD or CD bit unless DO is on, too */
+        if (!edns0_do) {
                 ad = false;
+                cd = false;
+
```
Comment (keszybz) [inline src/resolve/resolved-dns-stub.c:378]: Empty line.

**Review by keszybz — COMMENTED** [review]
OK, let's merge this. We still need to push the parts for resolvectl. All the patches here seem to work fine. Even if there's something to fix, we may just as well do this in later PRs.

**Pull request merged**